### ISSUE TITLE
Better Facts in E-Graphs

### ISSUE THREAD

**Issue by marcusrossel** [marcusrossel/lean-egg#49]
To handle rewrites with unbound conditions, we would like to use multi-patterns. For example, in a rewrite like:

```
?a => ?b if ?a - ?b = 0
```

... we can't apply the rewrite as is, as the pattern var in the LHS (`?a`) are not a subset of the pattern vars in the RHS (`?b`). With a mutli-pattern, we would match both `?a` and `?a - ?b = 0`, which would resolve `?b`, allowing us to apply the rewrite. Unfortunately, multi-patterns in egg break explanations (this simply isn't supported). To work around this, we need to find a way to use the LHS of the rewrite to represent both the term and the conditions we want to match.

### `With`-Nodes

For this, we introduce a `With(Id, Id)` constructor and write our conditional rewrite as follows:

```
(with ?a (?a - ?b = 0)) => ?b
```

For this to work as expected, there need to exist `with X Y` e-nodes in our e-graph for every possible e-class X and *known fact* Y (as we will see, this incurs an overhead of one e-node per e-class). To achieve this, we do two things:

1. We introduce an e-class (let's call it `F`) of all "known facts". That is, an e-class of all propositions which we know to be true, either because the user provided an explicit fact for this, or because we were able to derive this fact from other facts during eqsat.
2. We use the e-class analysis `modify` function to observe when new e-classes are created. If a new e-class `C` appears, we modify the e-graph by also adding the e-node `(with C F)`. Thus, any new e-class has a corresponding `with`-node. As the presence of `with`-nodes is invariant under e-class union, we obtain the invariant that every e-class has a corresponding `with`-node during the entire eqsat run.

### Fact E-Class

To implement Requirement (1) from above, we declare `F` to be the e-class of the term `const "True"`. Any propositional fact given by a user is added to `F`. To reify the equalities known by e-classes into explicit equality propositions in `F` we add a similar approach as used for Requirement (2) above. We use the `modify` function to check when a new e-class `C` is created, and in that case add a `(app (app (app (const "Eq" _) _) C) C)` e-node to `F`. That is, we reify the fact that any term in `C` is equal to any other term in `C`. Note the use of the term `_` which represents unknown subterms (as e.g. used for eta-expansion). We need to use `_` for the type and universe level of the equality, as we cannot easily reconstruct these. 

### Conjunction of Facts

To represent multiple conditions in a `with`-node, we also need to ensure that for any given facts `p` and `q`, the fact e-class `F` also contains `p ∧ q`. Ensuring this is trivial by adding the e-node `(app (app (const "And") F) F)` to the e-graph once before eqsat.

### Deduction of Facts

We can use the given setup to also deduce new facts from existing ones by applying given theorems. Namely, any given theorem of the form `∀ a b c h₁ h₂ ..., p` where `p` is not an equivalence can be turned into `∀ a b c h₁ h₂ ..., (p = True)`. From this, we can create a "normal" rewrite which needs to match (that is, prove) facts `h₁ h₂ ...` from the e-graph (if they are proven they are contained in `F`), and if successful rewrites `p => True` which adds `p` to the fact e-class `F`.

#### Open Problem:
If `p` is not yet contained in the e-graph, then the corresponding rewrite will never match. To partially resolve this, there's one possible adjustments I can think of: If the pattern vars in `p` are a subset of the pattern vars in `h₁ h₂ ...`, then we generate the rewrite as `(with True (h₁ ∧ h₂ ∧ ...))=> p`. When the pattern vars in `p` are not a subset of the pattern vars in `h₁ h₂ ...`, I'm not sure what to do. It would be nice to somehow add the unproven terms to the e-graph.

### Type Class Instances

As the fact e-class `F` merges all know facts into one e-class (that is, declares them to all be equivalent), we only want to add *propositional* facts into that e-class (as any two given proven propositions are provably equal). For other kinds of facts, namely the synthesizability of type class instances, we don't want to equate the underlying terms. Thus, type class instance conditions probably still have to be handled "the old way".

**Comment by memoryleak47** [comment]
> To represent multiple conditions in a with-node, we also need to ensure that for any given facts p and q, the fact e-class F also contains p ∧ q.

multiple conditions can also be simulated by multiple layers of "with".
So instead of `(with a (c1 ∧ c2))` we can do `(with (with a c1) c2)`.

**Comment by marcusrossel** [comment]
@memoryleak47 If we choose not to use `with`, then a nice effect of using $$\land$$ is that we only have 1 node overhead instead of one node per e-class (as with `with`).

**Comment by memoryleak47** [comment]
so in short, we require
- add `(fact True)` to the e-graph
- equate `(eq ?x ?x) = True` for each e-class `?x`
- equate `(and True True) = True`

and then we can express a rule `P1 -> P2 -> A=B`
by matching against `(fact (and P1 (and P2 (eq A A))))`,
and then applying the resulting subst to `A=B`.

Similarly, a "propositional" rule (i.e. a non-equality theorem / lemma from lean) P1 -> P2 -> P can be rewritten to an equational one, by `P1 -> P2 -> P=True`.

**Comment by memoryleak47** [comment]
> Open Problem: If p is not yet contained in the e-graph, then the corresponding rewrite will never match

Is `p` a condition, or the lhs of the equation? I would argue in both cases, it is intended that we cannot apply the rewrite rule. Maybe I'm misunderstanding

**Comment by memoryleak47** [comment]
> For other kinds of facts, namely the synthesizability of type class instances, we don't want to equate the underlying terms.

Why is that? I consider a type-class instance as a boolean term which is `True`, if Lean can synthesize a type-class and `False` otherwise.
To me it seems that the actual synthesized instance isn't really used, hence restricting it to the boolean question of "is this synthesizable?" is enough.
This probably again boils down to me not really understanding type-classes though ^^

**Comment by marcusrossel** [comment]
@memoryleak47 Agreed. That still leaves the question open of how we might introduce not--yet-proven propositions/ terms we might want to rewrite on into the e-graph in the first place. But that's somewhat orthogonal.

**Comment by marcusrossel** [comment]
This issue is now being worked on in the `better-facts` branch.

**Comment by marcusrossel** [comment]
> > Open Problem: If p is not yet contained in the e-graph, then the corresponding rewrite will never match
> 
> Is `p` a condition, or the lhs of the equation? I would argue in both cases, it is correct and intended that we cannot apply the rewrite rule.

If `p` is a (proven) proposition - that is, when packaged as a rewrite it would be `p => True` if `p` contains pattern variables, or `True => p` if it doesn't. And I agree with your point that it is correct and intended that we cannot apply the rewrite rule.

**Comment by marcusrossel** [comment]
Let's not do the `modify` approach, and add reified eq nodes in the runner hook instead.
